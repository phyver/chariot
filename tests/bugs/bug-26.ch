:set incremental_SCP false
:set depth 3
:set show_lists false
-- :set show_all_steps true

data nat where
    | Zero : nat
    | Succ : nat -> nat

val add m (n+1) = (add n m) + 1
  | add m 0 = m

data list('x) where
    | Nil : list('x)
    | Cons : 'x -> list('x) -> list('x)

val sum [] = 0
  | sum [n] = n
  | sum m::n::l = sum ((m+n)::l)

-- exemple 2
codata stream('x) where
    | Head : stream('x) -> 'x
    | Tail : stream('x) -> stream('x)

val sums { Head = [] ; Tail = s }      = { Head = 0 ; Tail = sums s }
  | sums { Head = [n] ; Tail = s }     = { Head = n ; Tail = sums s }
  | sums { Head = n::m::l ; Tail = s } = sums { Head = (add n m)::l ; Tail = s }
