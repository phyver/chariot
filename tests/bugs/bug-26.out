>> option verbose is set to 1
>> option incremental_SCP is set to false
>> option depth is set to 3
>> option show_lists is set to false
>> inductive type nat was succesfully defined
>> Typing for add successful
>> the definition for add is complete
>> the definition add is provably correct

>> inductive type list was succesfully defined
>> Typing for sum successful
>> the definition for sum is complete
>> the definition sum is provably correct

>> coinductive type stream was succesfully defined
>> Typing for sums successful
>> the definition for sums is complete
>> the definition sums is provably correct

